include/networkit/auxiliary/SparseVector.hpp

```diff
@@ -0,0 +1,229 @@
+/*
+ * SparseVector.hpp
+ *
+ * Created: 2019-10-15
+ * Author: Armin Wiebigke
+ */
+// networkit-format
+#ifndef NETWORKIT_AUXILIARY_SPARSE_VECTOR_HPP_
+#define NETWORKIT_AUXILIARY_SPARSE_VECTOR_HPP_
+
+#include <algorithm>
+#include <cassert>
+#include <vector>
+
+#include <networkit/Globals.hpp>
+
+namespace NetworKit {
+
+/**
+ * A vector that imitates a map with unsigned integer keys.
+ * This class has faster access than a map, but needs space linear in the maximum key value.
+ */
+template <typename T>
+class SparseVector {
```

Comment:
Does this data structure bring any advantage? As far as I see it does not save space or time.

Reply:
Save time or space compared to what other data structure? This data structure is a pattern we already use in NetworKit, e.g., in `PLM`, which could be refactored to use this new class. Compared to a hash map, it gives you guaranteed constant time operations and is also faster in practice. Note that it cannot always replace a hash map, only in cases where you can spend the time to initialize it once in time linear in the maximum key value, after which you get constant time insertion and querying of elements and reset linear in the number of inserted elements.

Reply:
Compared to a std vector. I was a bit confused by the name of this class, with `Sparse` I expected some sort of compression strategy like the one implemented a CSR matrix. This is not the case here but I see that `usedIndexes` is used to iterate over the "used" slots only, which is of course faster than iterating over all the elements if the elements stored are sparse.

Reply:
@angriman : What you describe is the compression strategy. You don't save space in this particular case, but time.